File: python/arrays_and_strings/is_anagram.py
"""
question:
Given two strings, write a function to check if strings are anagrams of each other. Anagrams are a strings that contains
the same characters in different order.
"""

def is_anagram(s1: str, s2: str) -> bool:
    """
    solution:
    check if two strings are anagrams. First make sure they are same length. Then count frequency of each character
    in first string and decrement count of each character in second string. If count of any character is zero before
    decrement, then its not an anagram. Store counts in hash map (dictionary) for fast lookup. Can initial a dictionary
    with entries for each letter in alphabet in constant time or add entry for first occurrence of each character in
    first string, and test for inclusion before decrementing count for each character in second string. Choose former
    b/c less verbose.

    Time: O(n)
    Space: O(1) - because finite number of letters in alphabet.
    """

    # if diff length, not anagrams
    if len(s1) != len(s2):
        return False

    # initialize dictionary with one entry for each letter of alphabet
    char_counts = {x: 0 for x in string.ascii_lowercase}

    # increment char counts for each char in s1
    for char in s1:
        char_counts[char.lower()] += 1

    # decrement char counts of each char in s2
    for char in s2:

        if char_counts[char.lower()] <= 0:
            # count is zero before decrement, so not anagrams
            return False
        else:
            char_counts[char.lower()] -= 1

    # they are anagrams!
    return True

File: javascript/arrays_and_strings/is_anagram.js
/*
question:
Given two strings, write a function to check if strings are anagrams of each other. Anagrams are a strings that contains
the same characters in different order.
*/

function isAnagram(s1, s2) {

  // not anagrams if not same length
  if (s1.length !== s2.length) return false

  // count letters in s1
  var counts = {}
  Array.prototype.forEach.call(s1, c => {
    if (c in counts) {
      counts[c] += 1;
    } else {
      counts[c] = 1
    }
  })

  // check s2
  Array.prototype.forEach.call(s2, c => {
    if (c in counts && counts[c] > 0) {
      counts[c] -= 1;
    } else {
      return false;
    }
  })

  // anagram if all counts at 0
  return !Object.keys(counts).map(c => counts[c]).some(x=>x>0)
}

// test

describe('is anagram', function() {
  it('returns true for two empty strings', function() {
    expect(isAnagram('','')).toEqual(true)
  })
  it('returns true if stings are anagrams of each other, false otherwise', function() {
    expect(isAnagram('a', 'a')).toEqual(true)
    expect(isAnagram('ab', 'ba')).toEqual(true)
    expect(isAnagram('ba', 'ba')).toEqual(true)
    expect(isAnagram('abcde', 'ecdab')).toEqual(true)
    expect(isAnagram('', 'b')).toEqual(false)
    expect(isAnagram('a', '')).toEqual(false)
    expect(isAnagram('a', 'b')).toEqual(false)
    expect(isAnagram('ab', 'b')).toEqual(false)
    expect(isAnagram('a', 'ba')).toEqual(false)
    expect(isAnagram('abcde', 'ecdaf')).toEqual(false)
  })
})
